File: go/problem30.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func main() {
	fmt.Println(SumDigitNPowers(3))
}

func SumDigitNPowers(n int) int {

	result := 0
	init := strings.Repeat("1", n-1)
	startIndex, err1 := strconv.Atoi(init)
	if err1 != nil {
		return -1
	}
	if startIndex == 1 {
		startIndex++
	}
	end := strings.Repeat("9", n+1)
	endIndex, err2 := strconv.Atoi(end)
	if err2 != nil {
		return -1
	}
	for i := startIndex; i <= endIndex; i++ {
		if i == sumPower(i, float64(n)) {
			result += i
		}
	}
	return result
}

func sumPower(num int, n float64) int {
	sum := 0.0
	for num > 0 {
		sum += math.Pow(float64(num%10), n)
		num = num / 10
	}
	return int(sum)
}

File: python/problem30.py
import math


def digitNPower(n):
    init = int("1" * (n-1))
    limit = int("9" * (n+1))
    if init == 1:
        init +=1
    result = 0
    for i in range(init, limit+1):
        sum = sumOfDigitsPowerN(i, n)
        if sum == i:
            result += i
    return int(result)


def sumOfDigitsPowerN(num, n):
    sum = 0
    original_num = num
    while (num > 0):
        sum += math.pow(num % 10, n)
        if sum > original_num:
            return sum
        num = num // 10
    return sum


if __name__ == "__main__":
    print(digitNPower(2))
